File: Bot/MiniMax.py
import Gordon
import random

class MiniMax(Gordon):
    min_eval_board = -1776
    max_eval_board = 1776

    def __init__(self, turn, depth):
        super().__init__(turn)
        self.depth = depth


    def eval_board(self, board, turn):
        temp_board = board.get_board()
        total_reward = 0

        for row in range(board.rows):
            for col in range(board.cols):
                if temp_board[row][col] == self.turn:
                    total_reward += self.VAL[row][col]
                if temp_board[row][col] == 3 - self.turn:
                    total_reward -= self.VAL[row][col]

        return total_reward


    def sort_nodes(self, moves):
        sorted_nodes = [[move, self.VAL[move[0]][move[1]]] for move in moves]

        sorted_nodes = sorted(sorted_nodes, key = lambda node: node[1], reverse = True)
        
        return [node[0] for node in sorted_nodes]


    def get_available_moves(self, board, player):
        board.set_available_moves(player)

        return super().get_available_moves(board)

    
    def minimax_alpha_beta(self, board, depth, player, Alpha, Beta, turn):
        valid_moves = self.get_available_moves(board, player)
        best_value = 0

        if depth == 0 or len(valid_moves) == 0:
            return self.eval_board(board, turn)

        if player == turn:
            tiles = sort_nodes(valid_moves)
            best_value = self.min_eval_board

            for piece in tiles:
                board_temp = copy.deepcopy(board)

                board_temp.make_move(player, piece[0], piece[1]) 
                val = self.minimax_alpha_beta(board_temp, depth - 1, 3 - player, Alpha, Beta, turn)
                
                #print(val, "Alpha_Beta_sorted() val maximize_player")

                if val > best_value:
                    best_value = val

                if val > Alpha:
                    Alpha = val

                if Alpha >= Beta:
                    break #beta cut-off
                
            #print(Alpha, "Alpha_Beta_sorted() Alpha maximize_player")
        else: # minimizingPlayer
            tiles = sort_nodes(valid_moves)
            best_value = self.max_eval_board

            for piece in tiles:
                board_temp = copy.deepcopy(board)

                board_temp.make_move(player, piece[0], piece[1])  
                val = self.minimax_alpha_beta(board_temp, depth - 1, 3 - player, Alpha, Beta, turn)
                
                #print(val, "Alpha_Beta_sorted() val minimize_player")

                if val < best_value:
                    best_value = val
                
                if val < Beta:
                    Beta = val

                if Alpha >= Beta:
                    break #alpha cut-off

            #print(Beta, "Alpha_Beta_sorted() Beta minimize_player")
        
        #print(best_value, "Alpha_Beta_sorted()", depth)
        return best_value


    def get_move(self, board):
        moves = self.get_available_moves(board, self.turn)
        sorted_tiles = sort_nodes(moves)

        max_points = self.min_eval_board
        best_move = []

        for move in sorted_tiles:
            board_temp = copy.deepcopy(board)
            board_temp.make_move(self.turn, move[0], move[1]) 

            points = self.minimax_alpha_beta(board_temp, self.depth, 3 - self.turn, max_points, self.max_eval_board, self.turn)
                    
            if points > max_points:
                max_points = points
                best_move = move
            elif points == max_points:
                if random.randint(0, 1):
                    best_move = move

        #print(best_move, "alpha_beta_sorted_move()")
        return best_move

File: demo/network/Game/MiniMax.cpp
#include "MiniMax.hpp"

const int MiniMax::REWARD[8][8] = {{120, -20, 20,  5,  5, 20, -20, 120},
                                   {-20, -40, -5, -5, -5, -5, -40, -20},
                                   { 20,  -5, 15,  3,  3, 15,  -5,  20},
                                   {  5,  -5,  3,  3,  3,  3,  -5,   5},
                                   {  5,  -5,  3,  3,  3,  3,  -5,   5},
                                   { 20,  -5, 15,  3,  3, 15,  -5,  20},
                                   {-20, -40, -5, -5, -5, -5, -40, -20},
                                   {120, -20, 20,  5,  5, 20, -20, 120}};
const int MiniMax::MIN_EVAL = -1776;
const int MiniMax::MAX_EVAL = 1776;

MiniMax::MiniMax(int turn, int depth): turn(turn), depth(depth) {}

int MiniMax::evalBoard(vector<vector<int>> board) {
    int totalReward = 0;

    for(int x = 0; x < board.size(); x++) {
        for(int y = 0; y < board[0].size(); y++) {
            if(board[x][y] == turn) {
                totalReward += REWARD[x][y];
            } else if(board[x][y] == 3 - turn) {
                totalReward -= REWARD[x][y];
            }
        }
    }

    return totalReward;
}

vector<vector<int>> MiniMax::getAvailableMoves(vector<vector<int>> board, int player) {
    Board temp_board = Board(false, board);
    temp_board.setAvailableMoves(turn);

    return Bot::getAvailableMoves(temp_board.getBoard());
}

int MiniMax::minimax(vector<vector<int>> board, int depth, int player) {
    vector<vector<int>> moves = getAvailableMoves(board, player);
    int bestValue;

    if(depth == 0 || moves.size() == 0) {
        return evalBoard(board);
    }

    if(player == turn) {
        bestValue = MIN_EVAL;

        for(int i = 0; i < moves.size(); i++) {
            Board temp_board = Board(false, board);

            temp_board.makeMove(player, moves[i][0], moves[0][1]);
            int val = minimax(temp_board.getBoard(), depth - 1, 3 - player);

            bestValue = val > bestValue ? val : bestValue;
        }
    } else {
        bestValue = MAX_EVAL;

        for(int i = 0; i < moves.size(); i++) {
            Board temp_board = Board(false, board);

            temp_board.makeMove(player, moves[i][0], moves[0][1]);
            int val = minimax(temp_board.getBoard(), depth - 1, 3 - player);

            bestValue = val < bestValue ? val : bestValue;
        }
    }

    return bestValue;
}

vector<int> MiniMax::makeMove(vector<vector<int>> board) {
    vector<vector<int>> moves = getAvailableMoves(board, turn);
    int maxPoints = MIN_EVAL;
    int choice;

    for(int i = 0; i < moves.size(); i++) {
        Board temp_board = Board(false, board);

        temp_board.makeMove(turn, moves[i][0], moves[0][1]);
        int val = minimax(temp_board.getBoard(), depth - 1, 3 - turn);

        if(val > maxPoints) {
            maxPoints = val;
            choice = i;
        }
    }

    //cout << "Hello!" << endl;
    return moves[choice];
}